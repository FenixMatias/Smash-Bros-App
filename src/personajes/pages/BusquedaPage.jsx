import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { PersonajesCard } from '../components';
import { getPersonajeByName } from '../helpers';

export const BusquedaPage = () => {

    const navigate = useNavigate();
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    const personajes = getPersonajeByName(q);
    const showSearch = (q.length === 0);
    const showError = (q.length > 0) && personajes.length === 0;

    const { busquedaText, onInputChange} = useForm({
        busquedaText: q
    });

    const onSearchSubmit = (event) => {
        event.preventDefault();

        // if(busquedaText.trim().length <= 1) return;

        navigate(`?q=${busquedaText.toLowerCase()}`)
    }

    return (
        <>
            <h1>Busqueda de Personajes</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Busqueda</h4>
                    <hr />
                    <form onSubmit={onSearchSubmit} role="searchForm">
                        <input 
                            type="text"
                            placeholder="Buscar Personaje"
                            className="form-control"
                            name="busquedaText"
                            autoComplete="off"
                            value={busquedaText}
                            onChange={onInputChange} 
                        />
                    <button className="btn btn-primary mt-2">
                        Buscar
                    </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Resultado</h4>
                    <hr />
                    {/* {
                        (q === '')
                        ? <div className="alert alert-primary">Personaje Encontrados</div>
                        : (personajes.length === 0) && <div className="alert alert-danger">No hay Personajes Encontrados con <b>{q}</b></div>
                    } */}
                    <div className="alert alert-primary animate__animated animate__fadeIn" style={{display: showSearch ? '' : 'none'}}>Buscar un Personaje</div>
                    <div aria-label="alert-danger" className="alert alert-danger animate__animated animate__fadeIn" style={{display: showError ? '' : 'none'}}>No hay Personajes Encontrados con <b>{q}</b></div>
                    
                    {
                        personajes.map(pj => (
                            <PersonajesCard key={pj.id}{...pj}/>
                        ))
                    }
                </div>
            </div>
            
        </>
    )
}