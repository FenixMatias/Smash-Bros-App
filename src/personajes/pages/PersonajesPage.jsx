import { useMemo } from 'react';
import { Navigate, useParams, useNavigate } from 'react-router-dom';
import { getPersonajeById } from '../helpers';

export const PersonajesPage = () => {

    const { id } = useParams();

    const personaje = useMemo( () => getPersonajeById(id), [id]); 

    const navigate = useNavigate();

    const onReturn = () => {

        navigate(-1);
    }

    if(!personaje){

        return <Navigate to="/64" />;
    }

    return (
        <div className="row mt-5 animate__animated animate__fadeInLeft">
            <div className="col-4">
                <img src={`/personajes/${id}.png`} 
                    alt={personaje.nombre} 
                    className="img-thumbnail"/>
            </div>
            <div className="col-8">
                <h3>{personaje.nombre}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><b>Compañia:</b> {personaje.compania}</li>
                    <li className="list-group-item"><b>Saga:</b> {personaje.saga}</li>
                    <li className="list-group-item"><b>Serie:</b> {personaje.serie}</li>
                    <li className="list-group-item"><b>Primera Aparición:</b> {personaje.primera_aparicion}</li>
                </ul>
                <button 
                    className="btn btn-info mt-3"
                    onClick={onReturn}>
                    Volver
                </button>
            </div>
        </div>
        
    )
}