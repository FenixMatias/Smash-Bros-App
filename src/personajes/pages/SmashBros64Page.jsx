import { PersonajesList } from '../components';


export const SmashBros64Page = () => {

    return (
       <>
            <h1>Super Smash Bros</h1>
            <hr />
            <PersonajesList serie='Super Smash Bros'/>
       </>
    )
}