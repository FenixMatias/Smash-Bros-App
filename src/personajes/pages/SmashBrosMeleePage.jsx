import { PersonajesList } from '../components';

export const SmashBrosMeleePage = () => {

    return (
        <>
            <h1>Super Smash Bros Melee</h1>
            <hr />
            <PersonajesList serie='Super Smash Bros Melee'/>
        </>
    )
}