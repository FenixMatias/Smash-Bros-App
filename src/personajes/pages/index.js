export * from './SmashBros64Page';
export * from './SmashBrosMeleePage';
export * from './PersonajesPage';
export * from './BusquedaPage';