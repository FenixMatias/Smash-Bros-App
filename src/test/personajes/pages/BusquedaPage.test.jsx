import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { BusquedaPage } from '../../../personajes/pages/BusquedaPage';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({

    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}));

describe('Pruebas en <BusquedaPage />', () => {

    beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrarse correctamente con valores por defecto ', () => {

        const {container} = render(
            <MemoryRouter>
                <BusquedaPage />
            </MemoryRouter>
        );

        // screen.debug()
        expect(container).toMatchSnapshot();
      
    });

    test('Debe de mostrar a Roy y el input con el valor del queryString ', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=roy']}>
                <BusquedaPage />
            </MemoryRouter>
        );

        const inputValue = screen.getByRole('textbox');

        expect(inputValue.value).toBe('roy');

        const img = screen.getByRole('img');

        expect(img.src).toBe('http://localhost/personajes/melee-roy.png');

        const alertDanger = screen.getByLabelText('alert-danger');

        expect(alertDanger.style.display).toBe('none');
      
    });

    test('Debe de mostrar un error si no encuentra un presonaje (roy123) ', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=roy123']}>
                <BusquedaPage />
            </MemoryRouter>
        );

        const alertDanger = screen.getByLabelText('alert-danger');

        expect(alertDanger.style.display).toBe('');
      
    });

    test('Debe de llamar el navigate a la pantalla nueva ()', () => {

        const inputValue = 'marth';

        render(
            <MemoryRouter initialEntries={['/search']}>
                <BusquedaPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        
        fireEvent.change(input, { target: { name: 'busquedaText', value: inputValue}});

        const form = screen.getByRole('searchForm');

        fireEvent.submit(form);

        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);

        // const searchButton = screen.getByRole('button', {name: 'Buscar'});

        // fireEvent.click(searchButton);

        // expect(mockUseNavigate).toHaveBeenCalledWith('?q=roy');
    
    });
    
});
