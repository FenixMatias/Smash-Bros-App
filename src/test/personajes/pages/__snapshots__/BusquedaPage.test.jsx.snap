// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <BusquedaPage /> Debe de mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Busqueda de Personajes
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Busqueda
      </h4>
      <hr />
      <form
        role="searchForm"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="busquedaText"
          placeholder="Buscar Personaje"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-2"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultado
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Buscar un Personaje
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hay Personajes Encontrados con 
        <b />
      </div>
    </div>
  </div>
</div>
`;
